overlay/tun_linux.go

```diff
@@ -265,22 +265,70 @@ func (t *tun) deviceBytes() (o [16]byte) {
 	return
 }
 
+func hasNetlinkAddr(al []*netlink.Addr, x netlink.Addr) bool {
+	for i := range al {
+		if al[i].Equal(x) {
+			return true
+		}
+	}
+	return false
+}
+
+// addIPs uses netlink to add all addresses that don't exist, then it removes ones that should not be there
+func (t *tun) addIPs(link netlink.Link) error {
+	newAddrs := make([]*netlink.Addr, len(t.vpnNetworks))
+	for i := range t.vpnNetworks {
+		//todo I wish I didn't need to stringify and re-parse
+		nlAddr, err := netlink.ParseAddr(t.vpnNetworks[i].String())
+		if err != nil {
+			return err
+		}
+		newAddrs[i] = nlAddr
+	}
+
+	//add all new addresses
+	for i := range newAddrs {
+		//todo do we want to stack errors and try as many ops as possible?
+		//todo AddrReplace should still add new IPs I think
+		if err := netlink.AddrReplace(link, newAddrs[i]); err != nil {
+			return err
+		}
+		//newAddrs is the same order as vpnNetworks so this is fine
+		if err := t.setDefaultRoute(t.vpnNetworks[i]); err != nil {
+			return err
+		}
+	}
+
+	//iterate over remainder, remove whoever shouldn't be there
+	al, err := netlink.AddrList(link, netlink.FAMILY_ALL)
+	if err != nil {
+		return fmt.Errorf("failed to get tun address list: %s", err)
+	}
+
+	for i := range al {
+		if hasNetlinkAddr(newAddrs, al[i]) {
+			continue
+		}
+		err = netlink.AddrDel(link, &al[i])
+		if err != nil {
+			t.l.WithError(err).Error("failed to remove address from tun address list")
+		} else {
+			t.l.WithField("removed", al[i].String()).Info("removed address not listed in cert(s)")
+		}
+	}
+
+	return nil
+}
+
 func (t *tun) Activate() error {
 	devName := t.deviceBytes()
 
 	if t.useSystemRoutes {
 		t.watchRoutes()
 	}
 
-	var addr, mask [4]byte
-
-	//TODO: IPV6-WORK
-	addr = t.cidr.Addr().As4()
-	tmask := net.CIDRMask(t.cidr.Bits(), 32)
-	copy(mask[:], tmask)
-
 	s, err := unix.Socket(
-		unix.AF_INET,
+		unix.AF_INET, //todo do we ever need INET6 here?
```

Comment:
in bsd land this needs to match the af of the routes you are trying to set, 6 for v6 and 4 for v4.

Reply:
In Linux-land we use netlink for that, and everything else is AF-independent. I left a comment in case that changes.